src/addons/metaservice/src/main/java/ee/ria/xroad/proxy/serverproxy/RestMetadataServiceHandlerImpl.java

```diff
@@ -0,0 +1,159 @@
+import ee.ria.xroad.common.util.MimeTypes;
+import ee.ria.xroad.common.util.MimeUtils;
+import ee.ria.xroad.proxy.protocol.ProxyMessage;
+import ee.ria.xroad.proxy.protocol.ProxyMessageDecoder;
+import ee.ria.xroad.proxy.protocol.ProxyMessageEncoder;
+
+import com.fasterxml.jackson.annotation.JsonInclude;
+import com.fasterxml.jackson.databind.ObjectMapper;
+import com.fasterxml.jackson.databind.PropertyNamingStrategy;
+import lombok.extern.slf4j.Slf4j;
+import org.apache.http.HttpStatus;
+import org.apache.http.client.HttpClient;
+import org.apache.http.impl.EnglishReasonPhraseCatalog;
+import org.apache.http.message.BasicHeader;
+
+import javax.servlet.http.HttpServletRequest;
+
+import java.util.Arrays;
+import java.util.Locale;
+
+import static ee.ria.xroad.common.metadata.MetadataRequests.ALLOWED_METHODS;
+import static ee.ria.xroad.common.metadata.MetadataRequests.LIST_METHODS;
+import static ee.ria.xroad.common.util.MimeUtils.HEADER_REQUEST_ID;
+
+/**
+ * Handler for REST metadata services
+ */
+@Slf4j
+public class RestMetadataServiceHandlerImpl implements RestServiceHandler {
+
+    static final ObjectFactory OBJECT_FACTORY = new ObjectFactory();
+
+    static final ObjectMapper MAPPER;
+
+    static {
+        final ObjectMapper mapper = new ObjectMapper();
+        mapper.setSerializationInclusion(JsonInclude.Include.NON_NULL);
+        mapper.setPropertyNamingStrategy(PropertyNamingStrategy.SNAKE_CASE);
+        MAPPER = mapper;
+    }
+
+    private RestResponse restResponse;
+    private CachingStream restResponseBody;
+
+    @Override
+    public boolean shouldVerifyAccess() {
+        return false;
+    }
+
+    @Override
+    public boolean shouldVerifySignature() {
+        return true;
+    }
+
+    @Override
+    public boolean shouldLogSignature() {
+        return true;
+    }
+
+    @Override
+    public boolean canHandle(ServiceId requestServiceId, ProxyMessage requestProxyMessage) {
+        if (requestProxyMessage.getRest().getVerb() != RestRequest.Verb.GET) {
+            return false;
+        }
+        return Arrays.asList(LIST_METHODS, ALLOWED_METHODS).contains(requestServiceId.getServiceCode());
+    }
+
+    @Override
+    public void startHandling(HttpServletRequest servletRequest, ProxyMessage requestProxyMessage,
+                              ProxyMessageDecoder messageDecoder, ProxyMessageEncoder messageEncoder,
+                              HttpClient restClient, HttpClient opMonitorClient,
+                              OpMonitoringData opMonitoringData) throws Exception {
+        restResponse = new RestResponse(requestProxyMessage.getRest().getClientId(),
+                requestProxyMessage.getRest().getQueryId(),
+                requestProxyMessage.getRest().getHash(),
+                requestProxyMessage.getRest().getServiceId(),
+                HttpStatus.SC_OK,
+                EnglishReasonPhraseCatalog.INSTANCE.getReason(HttpStatus.SC_OK, Locale.ENGLISH),
```

Comment:
Its a bit unclear if org.apache.http.**impl**.EnglishReasonPhraseCatalog is part of the public API. In addition, the reason phrase does not have semantics in HTTP protocol (https://tools.ietf.org/html/rfc7231#section-6.1), so I think it would be OK just to hard-code "OK" here, especially since there are no other responses.